python/related_numba.py

```diff
@@ -0,0 +1,90 @@
+from timing import lap, finish
+
+lap()
+import numpy as np
+import orjson
+from numba import njit, prange, typed
+
+
+@njit
+def get_all_top5(n_posts, t_to_pp, p_to_tt, p_to_nt):
+    all_top5 = np.empty((n_posts, 5), np.uint16)
+    for current_p in prange(n_posts):
+        tt = p_to_tt[current_p]
+        relation_count = np.zeros(n_posts, dtype=np.uint8)
+        for tj in range(p_to_nt[current_p]):
+            for p in t_to_pp[tt[tj]]:
+                relation_count[p] += 1
+        relation_count[current_p] = 0
+
+        # get top 5:
+        # (probably can be optimized more)
+        max_i = np.zeros(5, dtype=np.uint16)
+        max_r = np.zeros(5, dtype=np.uint8)
+        for i, r in enumerate(relation_count):
+            for j in range(4, -1, -1):
+                if r > max_r[j]:
+                    max_r[:j] = max_r[1:j + 1]
+                    max_i[:j] = max_i[1:j + 1]
+                    max_r[j] = r
+                    max_i[j] = i
+                    break
+        all_top5[current_p] = max_i
+    return all_top5
+
+
+def precompile():
+    lap()
+    # JIT compile by running with the arguments of the correct type
+    # 1) measure compile time
+    # 2) get correct processing (without compilation) time using the cached machine code
+    get_all_top5(
+        0,
+        typed.List([np.empty(0, dtype=np.uint16)]),
+        np.empty((0, 0), dtype=np.uint8),
+        np.empty(0, dtype=np.uint8),
+    )
+
+
+def main():
+    lap()
+    with open("../posts.json", "rb") as f:
+        s = f.read()
+        posts = orjson.loads(s)
+    lap()
+
+    unique_tags = set(tag for post in posts for tag in post["tags"])
```

Comment:
tags have to be represented as strings. Check the [rules](https://github.com/jinyus/related_post_gen#updated-results-from-github-workflow-raw-data)

Reply:
1. This change did not start in this PR. dd1f128be1e706aa9e28cefe5ab0619ec3894ca5
2. What is actually the problem? This is a set of strings.

Reply:
I meant this line:

```
tag_to_t = {t: np.uint8(i) for i, t in enumerate(unique_tags)}
```

Reply:
Impossible to use strings as array indices. Not in numba, not in numpy, not in C or Rust. Are you sure the rules are this strict: no numbering tags in any way?
Logically it's the same as having a matrix of size (tag_n x post_n) that has tag numbers as matrix indices. You will have to delete the numpy solution too. There is no workaround.

Reply:
Don't get me wrong, I will feel bad if both solutions are deleted, but if the aim of the repo is to benchmark just one exact algorithm written in different languages, then so be it. I think trying to trick slow languages into solving the problem quickly is fun, but this is not actually a clever solution competition and the rules not mine to write.

Reply:
The rule was made because everyone wanted to optimize away hashing and just use arrays. I think I will review the code and decide.

Reply:
I wonder if there's a way to tweak the problem so that hashing is actually optimal.